lib/http_v4_signature.c

```diff
@@ -0,0 +1,310 @@
+  char date[9];
+  char date_str[64];
+  const char *post_data = data->set.postfields ?
+    data->set.postfields : "";
+  const char *content_type = Curl_checkheaders(conn, "Content-Type");
+  unsigned char sha_d[32];
+  char sha_str[65];
+  char *cred_scope = NULL;
+  char *signed_headers = NULL;
+  char request_type[PROVIDER_MAX_L + sizeof("4_request")];
+  char *canonical_hdr = NULL;
+  char *canonical_request = NULL;
+  char *str_to_sign = NULL;
+  unsigned char tmp_sign0[32];
+  unsigned char tmp_sign1[32];
+  char *auth = NULL;
+  char *tmp;
+  int i;
+
+  (void) proxy;
+  if(Curl_checkheaders(conn, "Authorization")) {
+    /* Authorization already present, Bailing out */
+    return CURLE_OK;
+  }
+
+  if(!customrequest)
+    customrequest = "POST";
+  if(content_type) {
+    content_type = strchr(content_type, ':') + 1;
+    /* Skip whitespace now */
+    while(isblank(*content_type))
+      ++content_type;
+  }
+
+  /* Get Parameter
+     Google and Outscale use the same OSC or GOOG,
+     but Amazon use AWS and AMZ for header arguments */
+  tmp = strchr(provider, ':');
+  if(tmp) {
+    for(i = 0; provider != tmp; ++provider, ++i) {
+      if(i > PROVIDER_MAX_L)
+        return CURLE_FAILED_INIT;
+      up_provider[i] = (char)toupper(*provider);
+      low_provider0[i] = (char)tolower(*provider);
+    }
+    up_provider[i] = 0;
+    low_provider0[i] = 0;
+    ++provider;
+    for(i = 0; *provider; ++provider, ++i) {
+      if(i > PROVIDER_MAX_L)
+        return CURLE_FAILED_INIT;
+      low_provider[i] = (char)tolower(*provider);
+      mid_provider[i] = i ? low_provider[i] : (char)toupper(*provider);
+    }
+    low_provider[i] = 0;
+    mid_provider[i] = 0;
+  }
+  else if(strlen(provider) <= PROVIDER_MAX_L) {
+    for(i = 0; provider[i]; ++i) {
+      low_provider0[i] = (char)tolower(provider[i]);
+      low_provider[i] = (char)tolower(provider[i]);
+      up_provider[i] = (char)toupper(low_provider[i]);
+      mid_provider[i] = i ? low_provider[i] : up_provider[i];
+    }
+    low_provider0[i] = 0;
+    low_provider[i] = 0;
+    up_provider[i] = 0;
+    mid_provider[i] = 0;
+  }
+  else {
+    return CURLE_FAILED_INIT;
+  }
+
+  time(&rawtime);
+
+  ret = Curl_gmtime(rawtime, &info);
+  if(ret != CURLE_OK) {
+    return ret;
+  }
+
+  if(!strftime(date_iso, 256, "%Y%m%dT%H%M%SZ", &info)) {
+    return CURLE_OUT_OF_MEMORY;
+  }
+
+  memcpy(date, date_iso, 8);
+  date[8] = 0;
+  api_type = strdup(surl);
+  if(!api_type)
+    return CURLE_OUT_OF_MEMORY;
+
+  tmp = strchr(api_type, '.');
+  if(!tmp)
+    return CURLE_FAILED_INIT;
+  *tmp = 0;
+
+  region = strdup(strchr(surl, '.') + 1);
+  if(!region) {
+    ret = CURLE_OUT_OF_MEMORY;
+    goto free_all;
+  }
+
+  tmp = strchr(region, '.');
+  if(!tmp) {
+    ret = CURLE_FAILED_INIT;
+    goto free_all;
+  }
+  *tmp = 0;
+
+  uri = strchr(surl, '/');
+  host = strdup(surl);
+  if(!host) {
+    ret = CURLE_OUT_OF_MEMORY;
+    goto free_all;
+  }
+
+  query_url = strchr(host, '?');
```

Comment:
Don't get the query part of the URL like this it is error-prone, it is already parsed and extracted for use directly and correctly.